src/vmm/src/gdb/target.rs

```diff
@@ -0,0 +1,752 @@
+
+        let response = cpu_handle.response_receiver().recv()?;
+        if let VcpuResponse::NotAllowed(message) = response {
+            error!("Response from set regs: {message} on tid: {tid:?}");
+            return Err(Error::VcuRequestError);
+        }
+
+        Ok(())
+    }
+
+    /// Resumes the Vcpu, will return early if the Vcpu is already running
+    pub fn request_resume(&mut self, tid: Tid) -> Result<(), Error> {
+        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
+
+        if !vcpu_state.paused {
+            info!("Attempted to resume a vcpu already running.");
+            // Resuming an already running Vcpu is not considered an error case from GDB
+            return Ok(());
+        }
+
+        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
+
+        cpu_handle.send_event(VcpuEvent::Resume)?;
+
+        let response = cpu_handle.response_receiver().recv()?;
+        if let VcpuResponse::NotAllowed(message) = response {
+            error!("Response resume : {message}");
+            return Err(Error::VcuRequestError);
+        }
+
+        vcpu_state.paused = false;
+        Ok(())
+    }
+
+    /// Updates the kvm debug flags set against the Vcpu
+    fn update_kvm_debug(&self, tid: Tid) -> Result<(), Error> {
+        let vcpu_state = &self.vcpu_state[tid_to_vcpuid(tid)];
+        if !vcpu_state.paused {
+            info!("Attempted to update kvm debug on a non paused Vcpu");
+            return Ok(());
+        }
+
+        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
+
+        cpu_handle.send_event(VcpuEvent::SetKvmDebug(
+            self.hw_breakpoints.active_breakpoint(),
+            vcpu_state.single_step,
+        ))?;
+
+        match cpu_handle.response_receiver().recv() {
+            Ok(VcpuResponse::NotAllowed(message)) => {
+                error!("Response set kvm debug: {message}");
+                Err(Error::VcuRequestError)
+            }
+            Ok(VcpuResponse::Error(VcpuError::GdbRequest(e))) => Err(e),
+            Err(_) => Err(Error::VcuRequestError),
+            _ => Ok(()),
+        }
+    }
+
+    /// Translates from a guest virtual address to a physical address
+    fn translate_gva(&self, tid: Tid, address: u64) -> Result<u64, Error> {
+        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
+        cpu_handle.send_event(VcpuEvent::GvaTranslate(address))?;
+
+        let response = cpu_handle.response_receiver().recv()?;
+
+        if let VcpuResponse::GvaTranslation(response) = response {
+            return match response {
+                Some(res) => Ok(res.into()),
+                None => Err(Error::VcuRequestError),
+            };
+        }
+
+        if let VcpuResponse::NotAllowed(message) = response {
+            error!("Response from gva: {message}");
+            return Err(Error::VcuRequestError);
+        }
+
+        Ok(address)
+    }
```

Comment:
I think I see why we have this `Ok(address)` here: The way the handlers for `VcpuEvent::GvaTranslate` are set up, all possible responses are handled above, and this line is unreachable (but the compiler cannot know this). However, this means that if ever change the error handling over in vcpu/mod.rs, then we suddenly start returning something wrong here.

The more I try to find a suggestion of what to do instead, the more I realize how much of a struggle this channel based communication is, lol. I think the following should do the trick?

```suggestion
    fn translate_gva(&self, tid: Tid, address: u64) -> Result<u64, Error> {
        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
        cpu_handle.send_event(VcpuEvent::GvaTranslate(address))?;

        match cpu_handle.response_receiver().recv()? {
            VcpuResponse::GvaTranslation(translation) => Ok(translation?),
            other => Err(Error::UnexpectedResponse(other))
        }
    }
```

this assumes that `GvaTranslation` now holds a `Result`. 

Tbh, now that I'm looking at it, I really dislike how we're trying `VcpuResponse` as something that can hold both error and success variants. Imo it'd be so much cleaner if `cpu_handle.response_receiver().recv()` (which... why isn't it just `cpu_handle.receive_response()`, dual to `cpu_handle.send_event`? :disappointed:) would return a `Result<VcpuResponse, VcpuError>`, where now `VcpuResponse` only contains success variants, and `VcpuError` contains both variants for "something went wrong while processing the last `VcpuEvent`", as well as the error case from `.recv()`.

Reply:
With the design updates do you think we still want to use this change?